verilog/tools/ls/symbol-table-handler.cc

```diff
@@ -0,0 +1,249 @@
+      !status.ok()) {
+    LOG(WARNING) << "Could not find " << absl::GetFlag(FLAGS_file_list_path)
+                 << " file in the project:  " << status;
+    return;
+  }
+  LOG(INFO) << "Found file list under " << projectpath;
+  // fill the FileList object
+  if (absl::Status status = AppendFileListFromFile(projectpath, &filelist);
+      !status.ok()) {
+    // if failed to parse
+    LOG(WARNING) << "Failed to parse file list in " << projectpath << ":  "
+                 << status;
+    return;
+  }
+  // update include directories in project
+  for (const auto &incdir : filelist.preprocessing.include_dirs) {
+    LOG(INFO) << "Adding include path:  " << incdir;
+    currproject->AddIncludePath(incdir);
+  }
+  // add files from file list to the project
+  for (auto &incfile : filelist.file_paths) {
+    auto incsource = currproject->OpenIncludedFile(incfile);
+    if (!incsource.ok()) {
+      LOG(WARNING) << "File included in " << projectpath
+                   << " not found:  " << incfile << ":  " << incsource.status();
+      continue;
+    }
+    LOG(INFO) << "Creating symbol table for:  " << incfile;
+    BuildSymbolTableFor(*incsource.value());
+  }
+}
+
+bool IsStringViewContained(absl::string_view origin, absl::string_view substr) {
+  const int from = std::distance(origin.begin(), substr.begin());
+  const int to = std::distance(origin.begin(), substr.end());
+  if (from < 0) return false;
+  if (to > static_cast<int>(origin.length())) return false;
+  return true;
+}
+
+const SymbolTableNode *ScanReferenceComponents(
+    const ReferenceComponentNode *ref, absl::string_view symbol) {
+  if (IsStringViewContained(ref->Value().identifier, symbol))
+    return ref->Value().resolved_symbol;
+  for (const auto &childref : ref->Children()) {
+    const SymbolTableNode *resolved =
+        ScanReferenceComponents(&childref, symbol);
+    if (resolved) return resolved;
+  }
+  return nullptr;
+}
+
+const SymbolTableNode *SymbolTableHandler::ScanSymbolTreeForDefinition(
+    const SymbolTableNode *context, absl::string_view symbol) {
+  if (!context) {
+    return nullptr;
+  }
+  // TODO (glatosinski): reduce searched scope by utilizing information from
+  // syntax tree?
+  for (const auto &ref : context->Value().local_references_to_bind) {
+    if (ref.Empty()) continue;
+    const SymbolTableNode *resolved =
+        ScanReferenceComponents(ref.components.get(), symbol);
+    if (resolved) return resolved;
+  }
+  for (const auto &child : context->Children()) {
+    auto res = ScanSymbolTreeForDefinition(&child.second, symbol);
+    if (res) {
+      return res;
+    }
+  }
+  return nullptr;
+}
+
+std::vector<verible::lsp::Location> SymbolTableHandler::FindDefinition(
+    const verible::lsp::DefinitionParams &params,
+    const verilog::BufferTrackerContainer &parsed_buffers) {
+  const absl::Time finddefinition_start = absl::Now();
+  if (files_dirty_) {
+    BuildProjectSymbolTable();
```

Comment:
If we don't have a project, we should probably also finish here. So if `BuildProjectSymbolTable()` returns false (see comment above), then finish here with `return {};`